src/ncp/spinel_transformer.hpp

```diff
@@ -0,0 +1,35 @@
+/**
+ * NCP Spinel Transformer allows to transform spinel frames sent between Application Processor (AP) and Network Co-Processor (NCP).
+ */
+
+namespace SpinelTransformer {
+
+/**
+ * Transforms spinel frames before sending to AP/NCP.
+ *
+ * This method transforms outbound frames in both directions, i.e. from AP to NCP and from NCP to AP.
+ * NOTE: \p aOutputLength might be different than \p aInputLength, so make sure to allocate enough memory.
+ *
+ * @param[in] aInput Pointer to buffer containing spinel frame.
+ * @param[in] aInputLength Length of spinel frame.
+ * @param[out] aOutput Pointer to allocated buffer for transformed spinel frame.
+ * @param[in,out] aOutputLength Length of allocated buffer for transformed spinel frame.
+ * In return, it will contain the actual length of transformed frame.
+ * @return \c true on success, \c false otherwise.
+ */
+bool TransformOutbound(const unsigned char *aInput, size_t aInputLength, unsigned char *aOutput, size_t *aOutputLength);
+
+/**
+ * Restores spinel frames by transforming data received from AP/NCP.
+ *
+ * This method transforms inbound frames in both directions, i.e. from AP to NCP and from NCP to AP.
+ * @param[in] aInput Pointer to buffer containing received data.
+ * @param[in] aInputLength Length of received data.
+ * @param[out] aOutput Pointer to allocated buffer for restored spinel frame.
+ * @param[in,out] aOutputLength Length of allocated buffer for restored spinel frame.
+ * In return, it will contain the actual length of restored frame.
+ * @return \c true on success, \c false otherwise.
+ */
+bool TransformInbound(const unsigned char *aInput, size_t aInputLength, unsigned char *aOutput, size_t *aOutputLength);
```

Comment:
Would be reasonable to assume that `aInputLength` will always be larger than `*aOutputLength`? If that is always the case, then you could save a significant amount of memory (Getting rid of `NcpFrameBufferTransformerReader::mInputBuffer`) by requiring implementations of this method to support in-place transformation (i.e. where `aInput == aOutput`). This seems like it would be a reasonable requirement to place on a decryption implementation, since presumably the encrypted frame would have several bytes of header and if you are using something like AES-CCM you can decrypt on a byte-by-byte basis.

On that note, we could make `TransformOutbound` work similarly if we added a way for the caller to determine how large the header and footer of an encrypted frame would be. That would allow the caller to make sure that the plaintext frame is in the exact right spot for in-place encryption, allowing us to get rid of `NcpFrameBufferTransformerReader::mOutputBuffer` too.